### ISSUE TITLE
System.NullReferenceException when downloading Power Pages site

### ISSUE THREAD

**Issue by HirukaRanasinghe** [microsoft/powerplatform-build-tools#806]
I'm trying to download a Power Pages site that uses enhanced datamodel using the following command: 
`pac powerpages download --path [path] --webSiteId  [website ID] --modelVersion 2 -o`

However, I'm receiving the following error when it's downloading the content.

```
Information:
Be careful when you're updating public sites. The changes you make are visible to anyone immediately. To check site visibility go to https://make.powerpages.microsoft.com/.
To get additional information, please visit https://go.microsoft.com/fwlink/?linkid=2204350


Started downloading website [website ID]
Connected to... [Environment]
Connected as [user]
Downloading - webfiles
Downloading - powerpagesite
Downloading - powerpagesitelanguage
Downloading - powerpagecomponent
Downloaded - powerpagesitelanguage
Downloaded - powerpagesite
Downloaded - powerpagecomponent
Downloaded - webfiles
Sorry, the app encountered a non recoverable error and will need to terminate. The exception details have been captured and will be forwarded to the development team, if telemetry has been enabled. Exception Id: 2fadfdea-f1b5-4f6a-bd78-6bfe045d1163, Exception Type: System.NullReferenceException
The diagnostic logs can be found at: C:\**\pac-log.txt
``` 
My assumption is that there's something wrong with downloading mspp_* tables. How can I fix this error?

The log file has the following error: 
###   at  ***bolt.module.paportal.core.download.CoreEntityDownloader***.<>c__DisplayClass16_0.<GroupRecordsByComponentType>b__2(IGrouping`2** g) 

Here's an excerpt of the log file:
```
2024-04-08 04:31:42.8181 | FTL | bolt.Session :: Object reference not set to an instance of an object.|EXCEPTION: System.NullReferenceException: Object reference not set to an instance of an object.
   at bolt.module.paportal.core.download.CoreEntityDownloader.<>c__DisplayClass16_0.<GroupRecordsByComponentType>b__2(IGrouping`2 g)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at bolt.module.paportal.core.download.CoreEntityDownloader.GroupRecordsByComponentType(IDictionary`2 rawRecords, XmlDocument targetSchema)
   at bolt.module.paportal.core.download.CoreEntityDownloader.ExportDownloadedRecords(IDictionary`2 rawRecords, String tempRootFolder)
   at bolt.module.paportal.PortalDataExporter.ExportDataForCoreEntities(String exportFolder, CliExecutionContext cliInfra, DownloadContext downloadContext)
   at bolt.module.paportal.PortalDataExporter.ExportDataForAllEntities(String exportFolder)
   at bolt.module.paportal.PortalDataExporter.DownloadData()
   at bolt.module.paportal.verbs.PAPortalDownloadVerb.Execute(Command command)
   at bolt.cli.CommandLine.VerbExecutor.ExecuteImpl(VerbBase verbInstance, Command command)
   at bolt.cli.CommandLine.VerbExecutor.<>c__DisplayClass3_0.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.CommandLine.VerbExecutorBase.<ExecuteWithTelemetryAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.CommandLine.VerbExecutor.<ExecuteAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<>c__DisplayClass11_0.<<RunVerbExecutorAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<RunVerbExecutorAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<ExecuteWithOutputAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<ExecuteAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at bolt.cli.CommandProcessor.<ProcessCommands>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.CommandProcessor.<LaunchCommand>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.ModeManager.<Dispatch>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.Session.<Run>d__15.MoveNext()
2024-04-08 04:31:42.8181 | ERR | bolt.Session ::    at bolt.module.paportal.core.download.CoreEntityDownloader.<>c__DisplayClass16_0.<GroupRecordsByComponentType>b__2(IGrouping`2 g)
   at System.Linq.Enumerable.ToDictionary[TSource,TKey,TElement](IEnumerable`1 source, Func`2 keySelector, Func`2 elementSelector, IEqualityComparer`1 comparer)
   at bolt.module.paportal.core.download.CoreEntityDownloader.GroupRecordsByComponentType(IDictionary`2 rawRecords, XmlDocument targetSchema)
   at bolt.module.paportal.core.download.CoreEntityDownloader.ExportDownloadedRecords(IDictionary`2 rawRecords, String tempRootFolder)
   at bolt.module.paportal.PortalDataExporter.ExportDataForCoreEntities(String exportFolder, CliExecutionContext cliInfra, DownloadContext downloadContext)
   at bolt.module.paportal.PortalDataExporter.ExportDataForAllEntities(String exportFolder)
   at bolt.module.paportal.PortalDataExporter.DownloadData()
   at bolt.module.paportal.verbs.PAPortalDownloadVerb.Execute(Command command)
   at bolt.cli.CommandLine.VerbExecutor.ExecuteImpl(VerbBase verbInstance, Command command)
   at bolt.cli.CommandLine.VerbExecutor.<>c__DisplayClass3_0.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.CommandLine.VerbExecutorBase.<ExecuteWithTelemetryAsync>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.CommandLine.VerbExecutor.<ExecuteAsync>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<>c__DisplayClass11_0.<<RunVerbExecutorAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<RunVerbExecutorAsync>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<ExecuteWithOutputAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.Verb.<ExecuteAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at bolt.cli.CommandProcessor.<ProcessCommands>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.CommandProcessor.<LaunchCommand>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.cli.ModeManager.<Dispatch>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at bolt.Session.<Run>d__15.MoveNext()
2024-04-08 04:31:42.8258 | INF | bolt.Session :: End session. (sessionId: ########-####-####-####-############; ResponseCode: 500)

```

**Comment by yahiaouifaten** [comment]
I have the same issue.

**Comment by breves84** [comment]
Same issue. Site could previously be downloaded (as recent as February), but recent attempts to download the site have the same error as OP.

Another site in the same power platform environment (both sites on the enhanced data model) can download with no issues.

**Comment by jidmadhu** [comment]
We had same issue, but after removing the Card Layout from the Power Portal (Recent feature added into the power pages), everything seems working.

**Comment by D365Veteran** [comment]
> We had same issue, but after removing the Card Layout from the Power Portal (Recent feature added into the power pages), everything seems working.

Same issue here. How did you go about removing the Card Layout from the website?

**Comment by D365Veteran** [comment]
For anyone experiencing this bug with pac download. As mentioned above, it is due to the new UX Component for the Card Layout. To get around this, in dataverse search the Site Component (powerpagecomponent) table and filter on Component Type eq UX Component. From the content text you will be able to tell if the row is an entry for a Card Layout component. Deleting all Card Layout rows will resolve the issue.

**Comment by Bhawk90** [comment]
The issue is still happening and just to clarify, it is not only the Card Layout that is causing the issue. We've had the Breadcrumb UX Component for the Power Pages and that broke it as well. The log mentions "Entering method GroupRecordsByComponentType", I suppose they have forgot to add the UX Component component type for the grouping maybe?

**Comment by backiii** [comment]
Deleting the card layout rows also worked for me. But now having issues to update bootstrap to v5 with the enhanced data model. The migration starts and stops suddenly without an error message.

Issue is already described here.

https://github.com/microsoft/powerplatform-build-tools/issues/935

**Comment by TimVanOnckelen** [comment]
I'm having the same problem. It seems like adding codecompent tags are causing the issue.

**Comment by TimVanOnckelen** [comment]
> powerpagecomponent

This fixed to problem on my side.

**Comment by mannybellas** [comment]
I am also experience this issue (Version: 1.33.5+g1621296).  Workaround was to remove a "UX Component" type Site Component from the site.

**Comment by MrRishabhJain** [comment]
This issue has been fixed in Pac version 1.34.4 - https://www.nuget.org/packages/Microsoft.PowerApps.CLI/1.34.4
